tests/wot_coap_generation/generate.py

```diff
@@ -129,37 +134,151 @@ def extract_coap_resources(resources: list) -> list:
-    assert args.security, "ERROR: Argument security has to be defined"
-
-    thing_definiton = f"{current_directory}/config/wot_td/.thing.json"
+def get_wot_json(file: str, directory=THING_DESCRIPTION_DIRECTORY, validation_function=None) -> str:
+    path = f'{directory}/{file}'
     try:
-        f = open(thing_definiton)
-        thing_json = json.loads(f.read())
-        validate_thing_json(thing_json)
+        f = open(path)
+        wot_json = json.loads(f.read())
+        if validation_function is not None:
+            validation_function(wot_json)
     except IOError:
-        print(f"ERROR: Thing definition in {thing_definiton} is missing")
+        print(f"ERROR reading {path} is missing")
         sys.exit(0)
     except json.decoder.JSONDecodeError:
-        print(f"ERROR: json in {thing_definiton} is not valid")
+        print(f"ERROR: json in {path} is not valid")
         sys.exit(0)
     finally:
         f.close()
 
-    coap_affordances = f"{current_directory}/config/wot_td/.coap_affordances.json"
-    try:
-        f = open(coap_affordances)
-        coap_json = json.loads(
-            f.read(), object_pairs_hook=dict_raise_on_duplicates)
-        coap_jsons.append(coap_json)
-    except IOError:
-        print(f"INFO: Coap definition in {coap_affordances} not present")
-    except json.decoder.JSONDecodeError:
-        print(f"ERROR: json in {coap_affordances} is not valid")
-        sys.exit(0)
-    finally:
-        f.close()
-    generate_coap_resources()
+    return wot_json
+
+
+def parse_command_line_arguments() -> list:
+    parser = argparse.ArgumentParser(description='Web of Things helper script')
+    parser.add_argument('--board', help='Define used board')
+    parser.add_argument('--saul', action='store_true',
+                        help='Define if WoT TD SAUL is used')
+    parser.add_argument('--security', help='Define what security is used')
+    return parser.parse_args()
+
+
+def assert_command_line_arguments(args):
+    assert args.board, "ERROR: Argument board has to be defined"
+    assert args.security, "ERROR: Argument security has to be defined"
+
+
+def generate_includes(coap_resources: list) -> str:
+    dependencies = DEFAULT_DEPENDENCIES
+
+    dependencies = [f'#include {dependency}' for dependency in dependencies]
+    return "\n".join(dependencies)
+
+
+def generate_coap_listener() -> str:
+    result = f"static gcoap_listener_t {COAP_LISTENER_NAME} = {{\n"
+    result += INDENT + f"&{COAP_RESOURCES_NAME}[0],\n"
+    result += INDENT + f"ARRAY_SIZE({COAP_RESOURCES_NAME}),\n"
+    result += INDENT + f"{COAP_LINK_ENCODER_NAME},\n"
+    result += INDENT + "NULL,\n"
+    result += INDENT + "NULL\n"
+    result += "};"
+
+    return result
+
+
+def generate_coap_handlers(coap_resources: list) -> str:
+    handlers = []
+
+    for resource in coap_resources:
+        handler = f"static ssize_t {resource['handler']}(coap_pkt_t* pdu, uint8_t *buf, size_t len, void *ctx)\n"
```

Comment:
This is very useful for SAUL. You just need to place call the SAUL APIs within the function and return the values in standardized format. The form, data schema etc. should also get generated for it.